mlflow/tracing/trace_manager.py

```diff
@@ -0,0 +1,182 @@
+import logging
+import threading
+from contextvars import ContextVar
+from dataclasses import dataclass
+from functools import lru_cache
+from typing import Dict, Optional
+
+from opentelemetry import trace as trace_api
+
+from mlflow.tracing.types.model import Status, StatusCode, Trace, TraceData, TraceInfo
+from mlflow.tracing.types.wrapper import MLflowSpanWrapper, NoOpMLflowSpanWrapper
+
+_logger = logging.getLogger(__name__)
+
+
+# Internal representation to keep the state of a trace.
+# Dict[str, Span] is used instead of TraceData to allow access by span_id.
+@dataclass
+class _Trace:
+    trace_info: TraceInfo
+    span_dict: Dict[str, MLflowSpanWrapper]
+
+
+class InMemoryTraceManager:
+    """
+    Manage spans and traces created by the tracing system in memory.
+    """
+
+    _instance_lock = threading.Lock()
+    _instance = ContextVar("InMemoryTraceManager", default=None)
+
+    @classmethod
+    def get_instance(cls):
+        if cls._instance.get() is None:
+            with cls._instance_lock:
+                if cls._instance.get() is None:
```

Comment:
This check is for reducing the number of acquiring and releasing the lock to minimum. It shouldn't be visible overhead in normal case, but this object will be accessed quite frequently (basically whenever adding/updating spans), so I think it's safer to avoid unnecessary mutex operation:)